csrc/tensor_view.cpp

```diff
@@ -1379,6 +1379,9 @@ void TensorView::applyMmaSwizzle(MmaOptions options) {
   switch (options.operand) {
     case MmaOptions::Operand::Accumulator:
       mma_utils::WarpMmaSwizzler::scheduleMmaWarpOutput(this, options);
+      if (definition()->isA<MmaOp>()) {
+        setAllocationDomain(getLeafDomain(), true);
```

Comment:
:partying_face:

Reply:
Is this still non ideal? We shouldn't rely on leaf domain to dictate register layout, should we? My mental model for a leaf domain has been a loop nest. We shouldn't need to "loop index" into a register.

Reply:
There's nothing wrong to make the allocation layout the same as the loop structure. The leaf domain of a tensor does dictate its loop nest. In this case, it happens to be that its allocation layout is the same as the loop structure.

Reply:
I understood there's nothing wrong. If we look at the MMA instruction alone (`T9_l` in the following example), its leaf domain has the part `iMMA53{( ceilDiv(8, 4) )}, iMMA49{( ceilDiv(4, 2) )}, iMMA56{2}` that dictates how `T9_l` should index **into** the MMA register. If that part were in a different order (e.g. `iMMA56, iMMA53, iMMA49`) for whatever reasons, I believe that would be an invalid fusion IR. This leads me to think that, ideally, an MMA instruction should collapse **the part of** the leaf domain that dictates how it loop-indexes **into** the 2x2x2 MMA register, because the MMA instruction produces the register as a whole and can't control the loop nest **inside** the register. 

```
[ RUN      ] NVFuserTest.FusionVoltaMMATN_CUDA

%kernel {
T2_l[ iS4{16}, bS5{1}, iS6{4} ]
   = broadcast( T0_g[ iS0{16}, iS1{4} ] )
T5_s[ iS13{16}, bS14{1}, iS15{4} ]
   = Set( T2_l[ iS4{16}, bS5{1}, iS6{4} ], cache_op=Streaming )
T6_l[ ithreadIdx.x35{( ( ( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(1, 16) ) ) * 4 )}, bS32{8}, iS18{4} ]
   = Set( T5_s[ iS13{16}, bS14{1}, iS15{4} ], cache_op=Streaming )
T3_l[ bS7{1}, iS8{16}, iS9{4} ]
   = broadcast( T1_g[ iS2{16}, iS3{4} ] )
T7_s[ bS19{1}, iS20{16}, iS21{4} ]
   = Set( T3_l[ bS7{1}, iS8{16}, iS9{4} ], cache_op=Streaming )
T8_l[ bS36{( ceilDiv(1, 16) )}, bS39{8}, ithreadIdx.x46{( ( ( ( ceilDiv(8, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(16, 8) ) ) * 4 )}, iS24{4} ]
   = Set( T7_s[ bS19{1}, iS20{16}, iS21{4} ], cache_op=Streaming )
T9_l[ ithreadIdx.x59{( ( ( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(4, 2) ) ) * 2 )}, iMMA53{( ceilDiv(8, 4) )}, iMMA49{( ceilDiv(4, 2) )}, iMMA56{2}, rMMA12{4} ]
   = mma(T6_l[ ithreadIdx.x35{( ( ( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(1, 16) ) ) * 4 )}, bS32{8}, iS18{4} ],
         T8_l[ bS36{( ceilDiv(1, 16) )}, bS39{8}, ithreadIdx.x46{( ( ( ( ceilDiv(8, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(16, 8) ) ) * 4 )}, iS24{4} ])
T4_g[ ithreadIdx.x72{( ( ( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(4, 2) ) ) * 2 )}, iS66{( ceilDiv(8, 4) )}, iS62{( ceilDiv(4, 2) )}, iS69{2} ]
   = Set( T9_l[ ithreadIdx.x59{( ( ( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(4, 2) ) ) * 2 )}, iMMA53{( ceilDiv(8, 4) )}, iMMA49{( ceilDiv(4, 2) )}, iMMA56{2}, rMMA12{4} ], cache_op=Streaming )
```

```
T9_l[ ithreadIdx.x59{( ( ( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(4, 2) ) ) * 2 )}, iMMA53{( ceilDiv(8, 4) )}, iMMA49{( ceilDiv(4, 2) )}, iMMA56{2}, rMMA12{4} ]
 root domain : (iS10{16}, iS11{16}, rMMA12{4})
 allocation domain : (ithreadIdx.x59{( ( ( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(4, 2) ) ) * 2 )}, iMMA53{( ceilDiv(8, 4) )}, iMMA49{( ceilDiv(4, 2) )}, iMMA56{2}, rMMA12{4})
 contiguity: t t t t n
  Split: iS10{16} by factor 4 -> iS47{( ceilDiv(16, 4) )}, iS48{4}, start offset: 0, stop offset: 0
  Split: iS11{16} by factor 8 -> iS51{( ceilDiv(16, 8) )}, iS52{8}, start offset: 0, stop offset: 0
  Merge: iS47{( ceilDiv(16, 4) )} and iS51{( ceilDiv(16, 8) )} -> iS57{( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) )}
  Split: iS52{8} by factor 4 -> iMMA53{( ceilDiv(8, 4) )}, iS54{4}, start offset: 0, stop offset: 0
  Split: iS54{4} by factor 2 -> iS55{( ceilDiv(4, 2) )}, iMMA56{2}, start offset: 0, stop offset: 0
  Merge: iS57{( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) )} and iS55{( ceilDiv(4, 2) )} -> iS58{( ( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(4, 2) ) )}
  Split: iS48{4} by factor 2 -> iMMA49{( ceilDiv(4, 2) )}, iS50{2}, start offset: 0, stop offset: 0
  Merge: iS58{( ( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(4, 2) ) )} and iS50{2} -> ithreadIdx.x59{( ( ( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(4, 2) ) ) * 2 )}
 leaf domain : (ithreadIdx.x59{( ( ( ( ceilDiv(16, 4) ) * ( ceilDiv(16, 8) ) ) * ( ceilDiv(4, 2) ) ) * 2 )}, iMMA53{( ceilDiv(8, 4) )}, iMMA49{( ceilDiv(4, 2) )}, iMMA56{2}, rMMA12{4})
```

Reply:
@wujingyue Are you talking about merging `iMMA53{( ceilDiv(8, 4) )}, iMMA49{( ceilDiv(4, 2) )}, iMMA56{2}` into a single ID during schedule? Conceptually this does make more sense to me, but in practice, this makes no difference but, because loops parallelized as `MMA` is by definition "trivial", therefore will never have index.

Reply:
> We shouldn't rely on leaf domain to dictate register layout, should we?

I think it is the opposite. In general, we should use leaf domain as leaf domain as much as possible. For example, if I have
```
T1 = T0 (input);
T2 (output) = T1;
```
Assume that the leaf domain for `T1` and `T2` are both `[BIDx, TIDx, 8]`, then when indexing `T1` as a consumer, because the leaf domain is already allocation domain, so the indexing is trivial: the index is just the loop variable for the dimension of size 8. If instead, we use rFactor domain as allocation, we need to traverse the transformation from rFactor to leaf to back propagate the index. When indexing `T1` as a producer, we need to replay `T1` as `T2`, because the leaf domain of `T2` is already identical to the allocation domain of `T1`, we can then just trivially use the loop index again as the index of `T1`.

Reply:
I agree with you, but one key question remains not answered: I understand `T9_l`'s leaf domain is currently `ithreadIdx.x59, iMMA53, iMMA49, iMMA56, rMMA12`. If it were instead `ithreadIdx.x59, iMMA56, iMMA53, iMMA49, rMMA12` instead, should codegen (1) error out or (2) produce the same kernel or (2) produce a functional but slower kernel? Note that I'm asking what it **should** do (the contract), not what it *does* (the implementation). Understanding the contract helps me program with the right mental model.

Reply:
A side question: `rMMA12{4}` in `T9_l`'s root/allocation/leaf domain means a reduction? I don't see reduction ops around it. Does it bear a special meaning?

Reply:
> codegen (1) error out or (2) produce the same kernel or (2) produce a functional but slower kernel?

Ideally, it should error out, but unfortunately there is no good way to reliablely detect errors without duplicating the implementation another time. Generally, I would like to consider our lowering an a programming language like C++, and our schedulers as some automated C++ code generator. In this mindset, then the mma schedules are inline assembly `asm` to C++. Ideally, C++ compilers should check the correctness of inlined assmebly, but in practice, it is not checked. Our hardware just definitely this special weird memory format for mma, and we should schedule consistent with it. If not, the behavior is undefined. It can be silent wrong result, it can also be misaligned address access.

Reply:
> A side question: `rMMA12{4}` in `T9_l`'s root/allocation/leaf domain means a reduction? I don't see reduction ops around it. Does it bear a special meaning?

Yes, it is a reduction. Mma operators is a fused mul + sum.